lib/ws_deque.ml

```diff
@@ -0,0 +1,196 @@
+(*
+ * Copyright (c) 2015, Théo Laurent <theo.laurent@ens.fr>
+ * Copyright (c) 2015, KC Sivaramakrishnan <sk826@cl.cam.ac.uk>
+ * Copyright (c) 2017, Nicolas ASSOUAD <nicolas.assouad@ens.fr>
+ * Copyright (c) 2021, Tom Kelly <ctk21@cl.cam.ac.uk>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(* Work Stealing Queue
+ *
+ * See:
+ *   Dynamic circular work-stealing deque
+ *   https://dl.acm.org/doi/10.1145/1073970.1073974
+ *  &
+ *   Correct and efficient work-stealing for weak memory models
+ *   https://dl.acm.org/doi/abs/10.1145/2442516.2442524
+ *)
+
+module type S = sig
+  type 'a t
+  val create : unit -> 'a t
+  val is_empty : 'a t -> bool
+  val size : 'a t -> int
+  val push : 'a t -> 'a -> unit
+  val pop : 'a t -> 'a
+  val steal : 'a t -> 'a
+end
+
+module CArray = struct
+
+  type 'a t = {
+    arr  : 'a array;
+    mask : int
+    }
+
+  let rec log2 n =
+    if n <= 1 then 0 else 1 + (log2 (n asr 1))
+
+  let create n v =
+    let sz = Int.shift_left 1 (log2 n) in
+    assert ((sz >= n) && (sz > 0));
+    assert (Int.logand sz (sz-1) == 0);
+    {
+      arr  = Array.init sz (fun _ -> v);
+      mask = sz - 1
+    }
+
+  let size t = Array.length t.arr [@@inline]
+
+  let get t i =
+    Array.unsafe_get t.arr (Int.logand i t.mask) [@@inline]
+
+  let put t i v =
+    Array.unsafe_set t.arr (Int.logand i t.mask) v [@@inline]
+
+  let grow t top bottom =
+    let s = size t in
+    let ns = 2 * s in
+    let out = create ns (Obj.magic ()) in
+    for i = top to bottom do
```

Comment:
There is https://ocaml.org/api/Array.html#VALblit which may or may not be faster given that `get` and `put` do not perform bounds checks.